backends/itt/gen_itt_library.rb

```diff
@@ -0,0 +1,337 @@
+require_relative 'gen_itt_library_base.rb'
+
+def unwrap_typedef_to_concrete(t)
+  # unwrap chains like Typedef -> Typedef -> Enum
+  seen = {}
+  while t.respond_to?(:type) && !seen[t.object_id]
+    seen[t.object_id] = true
+    t = t.type
+  end
+  t
+end
+
+def enum_with_members_or_nil(e)
+  e if e && e.respond_to?(:members) && e.members && !e.members.empty?
+end
+
+def find_enum_by_name(name, api)
+  base = name.to_s
+  base_stripped = base.sub(/_t\z/, '')
+
+  enums = Array(api["enums"])
+  typedefs = Array(api["typedefs"])
+
+  # direct match (array or hash)
+  if api["enums"].is_a?(Hash)
+    e = api["enums"][base] || api["enums"][base_stripped]
+    return e if enum_with_members_or_nil(e)
+  else
+    e = enums.find { |x| x.respond_to?(:name) && (x.name == base || x.name == base_stripped) }
+    return e if enum_with_members_or_nil(e)
+  end
+
+  # typedef: name or stripped name
+  td = typedefs.find { |t| t.respond_to?(:name) && (t.name == base || t.name == base_stripped) }
+  if td
+    e = unwrap_typedef_to_concrete(td)
+    return e if enum_with_members_or_nil(e)
+  end
+
+  # forward-declared enum object; look for another enum with same name but members set
+  if e && e.respond_to?(:name)
+    alt = enums.find { |x| x != e && x.respond_to?(:name) && x.name == e.name && enum_with_members_or_nil(x) }
+    return alt if alt
+  end
+
+  # look for any anonymous enum that has members
+  if td
+    anon = unwrap_typedef_to_concrete(td)
+    return anon if enum_with_members_or_nil(anon)
+  end
+
+  nil
+end
+
+def print_enum(name, enum)
+  by_name = {}
+  current = -1
+
+  resolve_alias = ->(ref) do
+    by_name.fetch(ref.to_sym) { raise "enum #{name}: alias #{ref} not defined yet" }
+  end
+
+  parse_string = ->(s) do
+    s = s.strip
+    return resolve_alias.call(s[1..]) if s.start_with?(':')               # Ruby-style symbol alias
+    return resolve_alias.call(s)      if /\A[_A-Za-z]\w*\z/.match?(s)     # C-style ident alias
+    # Last resort: numeric/expr
+    Integer(s)
+  rescue ArgumentError
+    eval(s)
+  end
+
+  resolved = enum.members.map do |m|
+    n = m.name.to_sym
+    v = m.val
+
+    current =
+      case v
+      when nil       then current + 1
+      when Integer   then v
+      when Symbol    then resolve_alias.call(v)
+      when String    then parse_string.call(v)
+      else
+        raise "enum #{name}: unsupported val #{v.inspect} for #{n}"
+      end
+
+    by_name[n] = current
+    [n, current]
+  end
+
+  pair = ->(sym_val) { sym, val = sym_val; "#{sym.inspect}, #{val}" }
+
+  puts <<~RUBY
+    #{to_class_name(name)} = ittenum #{to_ffi_name(name)},
+      [ #{resolved.map(&pair).join(",\n        ")} ]
+  RUBY
+end
+
+def print_itt_object(object)
```

Comment:
`print_<be>_object` does the same thing as `print_pointer_type`. Why have a specialization here instead of removing the code and doing `close_type` in the pointer part of the big if...else section (line 321)?

Reply:
Cool, I figured there was some history behind it. I'll leave it as is for now to mirror the other backends (and be cleaned up in a future refactor).